algorithms/src/std_algorithms/numeric/Kokkos_AdjacentDifference.hpp

```diff
@@ -0,0 +1,295 @@
+struct StdAdjacentDifferenceDefaultBinaryOpFunctor {
+  KOKKOS_FUNCTION
+  RetType operator()(const ValueType1& a, const ValueType2& b) const {
+    return a - b;
+  }
+};
+
+template <class IteratorType, class DestViewType, class BinaryOperator>
+struct StdAdjacentDiffItToViewFunctor {
+  using index_type = typename IteratorType::difference_type;
+
+  const IteratorType m_first;
+  const DestViewType m_dest_view;
+  BinaryOperator m_op;
+
+  KOKKOS_FUNCTION
+  void operator()(const index_type i) const {
+    const auto& my_value = m_first[i];
+    if (i == 0) {
+      m_dest_view(i) = my_value;
+    } else {
+      const auto& left_value = m_first[i - 1];
+      m_dest_view(i)         = m_op(my_value, left_value);
+    }
+  }
+
+  KOKKOS_FUNCTION
+  StdAdjacentDiffItToViewFunctor(IteratorType first, DestViewType dest_view,
+                                 BinaryOperator op)
+      : m_first(first), m_dest_view(dest_view), m_op(op) {}
+};
+
+template <class ViewTypeFrom, class OutputIteratorType>
+struct StdAdjDiffCopyFunctor {
+  ViewTypeFrom m_view_from;
+  OutputIteratorType m_first_dest;
+
+  KOKKOS_FUNCTION
+  StdAdjDiffCopyFunctor(const ViewTypeFrom view_from,
+                        OutputIteratorType first_dest)
+      : m_view_from(view_from), m_first_dest(first_dest) {}
+
+  KOKKOS_FUNCTION
+  void operator()(int i) const { m_first_dest[i] = m_view_from(i); }
+};
+
+// ------------------------------------------
+// adjacent_difference_impl
+// ------------------------------------------
+template <class ExecutionSpace, class InputIteratorType,
+          class OutputIteratorType, class BinaryOp>
+OutputIteratorType adjacent_difference_impl(const std::string& label,
+                                            const ExecutionSpace& ex,
+                                            InputIteratorType first_from,
+                                            InputIteratorType last_from,
+                                            OutputIteratorType first_dest,
+                                            BinaryOp bin_op) {
+  // checks
+  static_assert_random_access_and_accessible(ex, first_from, first_dest);
+  static_assert_iterators_have_matching_difference_type(first_from, first_dest);
+  expect_valid_range(first_from, last_from);
+
+  if (first_from == last_from) {
+    return first_dest;
+  }
+
+  // potentially, we could handle better the case where
+  // source interval and output interval are not the same.
+  // in that case, we don't need the auxiliary view.
+
+  // aliases
+  using value_type    = typename OutputIteratorType::value_type;
+  using aux_view_type = ::Kokkos::View<value_type*, ExecutionSpace>;
+  using functor1_t    = StdAdjacentDiffItToViewFunctor<InputIteratorType,
+                                                    aux_view_type, BinaryOp>;
+  using functor2_t = StdAdjDiffCopyFunctor<aux_view_type, OutputIteratorType>;
+
+  // run
+  const auto num_elements = last_from - first_from;
```

Comment:
we in fact have a static assertion in every function overload accepting iterators that they must be random access :) 
```cpp
  // checks
  Impl::static_assert_random_access_and_accessible(ex, first);
```
But yes, we could change the `const auto num_elements = last_from - first_from;` to use `distance` since we implement that too

Reply:
Replaced with `Kokkos::Experimental::distance`.

Reply:
Note: `Kokkos::Experimental::distance` also static-asserts that the iterators are random access.

Now I guess the question remains whether we should use `InputIterator` in things like:
```cpp
template <class ExecutionSpace, class InputIterator, class T>
InputIterator find(...);
```